### PR TITLE
Statistics

### PR DESCRIPTION
**Pull request by FredrikBixo**
What's new in this pull request:

- Added statistics for the user object
- Add support for uploading and downloading stats to a user via firebase.
- Added textfields in the profile menu for stats.

TODO: Verification + validation. Does it fullfil the requirements

### CODE REVIEW

File: app/src/main/res/layout/activity_profile_page.xml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    <android.support.design.widget.FloatingActionButton
+        android:id="@+id/fab"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="bottom|end"
+        android:layout_margin="@dimen/fab_margin"
+        app:srcCompat="@android:drawable/ic_dialog_email" />
+
+</android.support.design.widget.CoordinatorLayout>
```
Comment (LudoHoff) [inline app/src/main/res/layout/activity_profile_page.xml]: What is this profile page ? Shouldn't it be with the pop up layout ?

File: app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/CurrentUser.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.google.firebase.auth.FirebaseAuth;
+
+public class CurrentUser extends User {
+
+    CurrentUser() {
+        super(FirebaseAuth.getInstance().getCurrentUser().getUid());
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (FredrikBixo) [reply app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/CurrentUser.java]: This class is not used yet, but it's used to obtain the current user using the device.

File: app/src/main/res/layout/activity_profile_page.xml (hunk 1 above)
Comment (FredrikBixo) [reply app/src/main/res/layout/activity_profile_page.xml]: Yes, you are right. It's not used. This class can be deleted

File: app/src/androidTest/java/ch/epfl/sweng/SDP/Matchmaking/MatchmakerTest.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        Matchmaker.getInstance().joinRoom();
+        assertTrue(true);
+
+    }
+
+    @Test
+    public void leaveRoom() {
+        Matchmaker.getInstance().leaveRoom("2312");
+        assertTrue(true);
+    }
```
[the remark this answers is not in the document]
Comment (FredrikBixo) [reply app/src/androidTest/java/ch/epfl/sweng/SDP/Matchmaking/MatchmakerTest.java]: I'm not sure how I should test the firebase functions. Do you have any idea?

File: app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/Matchmaker.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+     * leave a room.
+     * @param roomId the id of the room.
+     */
+    public void leaveRoom(String roomId) {
+        // FirebaseUser currentFirebaseUser = FirebaseAuth.getInstance().getCurrentUser() ;
+        myRef.child(roomId).child("users").child("123").removeValue();
+    }
+
+
+}
```
[the remark this answers is not in the document]
Comment (FredrikBixo) [reply app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/Matchmaker.java]: It's the same branch

File: app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/User.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                // Update room
+
+                name = (String) map.get("name");
+                name = (String) map.get("username");
+                id = (String) map.get("id");
+
+                stars = toIntExact( (Long) map.get("stars"));
+                trophies = toIntExact((Long) map.get("trophies"));
+                matchesWon = toIntExact( (Long) map.get("matchesWon"));
+                matchesLost = toIntExact((Long) map.get("matchesLost"));
```
Comment (rbischof) [inline app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/User.java]: I don't find it necessary to save matchesLost in Database. You could use matchesPlayed - matchesWon.

File: app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/User.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+                matchesPlayed = toIntExact((Long) map.get("matchesPlayed"));
+                averageRating = toIntExact((Long) map.get("averageRating"));
```
Comment (rbischof) [inline app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/User.java]: How are you going to calculate the average rating when a new rating is available? ((averageRating*matchesPlayed)+newRating)/(matchesPlayed+1)? Is a bit complicated.
Maybe save totalRating in database and display it like this: totalRanking/matchesPlayed

File: app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/User.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+import com.google.firebase.database.FirebaseDatabase;
+import com.google.firebase.database.ValueEventListener;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import static java.lang.Math.toIntExact;
+
+public class User implements Serializable {
```
Comment (rbischof) [inline app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/User.java]: We already have a class called Account which implements most of these functions. Andrea is creating a locally cached instance of the user account where you can retrieve the stars and trophies from. Talk to him to also store matches played/won.

File: functions/index.js [hunk 8]
```diff
@@ -110,9 +113,11 @@ exports.joinGame = functions.https.onRequest((req, res) => {
         console.log('rooms[room].users is undefined');
       }
     }
+    var o = {  };
+    o[userId] =  { name : "fredrik"};
```
Comment (rbischof) [inline functions/index.js]: This should probably not be hardcoded to "fredrick".

File: app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/Matchmaker.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                }
+            }
+    }
+
+    /**
+     * leave a room.
+     * @param roomId the id of the room.
+     */
+    public void leaveRoom(String roomId) {
+        myRef.child(roomId).child("users").child("123").removeValue();
```
Comment (rbischof) [inline app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/Matchmaker.java]: This should probably not be hardcoded to 123?

File: functions/index.js [hunk 10]
```diff
@@ -99,9 +100,11 @@ exports.joinGame = functions.https.onRequest((req, res) => {
               // set playing to true
               admin.database().ref('rooms').child(room).child('playing').set(true)
             }
-            return  admin.database().ref('rooms').child(room).child('users').push().set({
+            // Push new user
+            return  admin.database().ref('rooms').child(room).child('users').child(userId).set({
               name: "fredrik"
```
Comment (rbischof) [inline functions/index.js]: This should probably not be hardcoded to fredrick.

File: app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/Matchmaker.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        return singleInstance;
+    }
+
+
+    /**
+     *  Matchmaker init.
+     */
+    public Matchmaker() {
```
Comment (rbischof) [inline app/src/main/java/ch/epfl/sweng/SDP/Matchmaking/Matchmaker.java]: I'm not sure, but shouldn't the constructor of a singleton instance be private?

File: app/src/main/java/ch/epfl/sweng/SDP/home/HomeActivity.java [hunk 12]
```diff
@@ -241,8 +244,23 @@ private void pressButton(View view) {
...
+        wonTextView.setText("Matches won: " + auser.getMatchesWon());
+
+        TextView lostTextView = profileWindow.findViewById(R.id.game_lost);
+        lostTextView.setText("Matches lost: " + auser.getMatchesLost());
+
+        TextView avgTextView = profileWindow.findViewById(R.id.averageStars);
+        avgTextView.setText("Avrage stars: "  + auser.getAverageRating());
+
+        TextView trophiesTextView = profileWindow.findViewById(R.id.maxTrophies);
+        trophiesTextView.setText("Avrage trophies: "  + auser.getTrophies());
```
[the remark this answers is not in the document]
Comment (FredrikBixo) [reply app/src/main/java/ch/epfl/sweng/SDP/home/HomeActivity.java]: Yes, getTrophies is returns maxTrophies

**Pull request merged**